### PR TITLE
SONARJAVA-3287 S5122 should only raise when all origins are allowed

### PR DESCRIPTION
**Pull request by quentin-jaquier-sonarsource**

### CODE REVIEW

File: java-checks/src/main/java/org/sonar/java/checks/CORSCheck.java [hunk 1]
```diff
@@ -90,24 +84,46 @@ private void checkMethod(Tree tree) {
...
-        String constantCORS = ExpressionsHelper.getConstantValueAsString(mit.arguments().get(0)).value();
-        if (constantCORS != null && HTTP_HEADERS.contains(constantCORS.toLowerCase(Locale.ENGLISH))) {
+        String headerName = ExpressionsHelper.getConstantValueAsString(mit.arguments().get(0)).value();
+        if (ACCESS_CONTROL_ALLOW_ORIGIN.equalsIgnoreCase(headerName) && isStar(mit.arguments().get(1))) {
           reportTree(mit.methodSelect());
         }
       } else if (APPLY_PERMIT_DEFAULT_VALUES.matches(mit)) {
         applyPermit.add(mit);
-      } else if (ADD_ALLOWED_ORIGIN.matches(mit)) {
+      } else if (ADD_ALLOWED_ORIGIN.matches(mit) && isStar(mit.arguments().get(0))) {
```
Comment (quentin-jaquier-sonarsource) [inline java-checks/src/main/java/org/sonar/java/checks/CORSCheck.java]: This is to support the case:
> CorsConfiguration: raises only when addAllowedOrigin("*")

File: java-checks/src/main/java/org/sonar/java/checks/CORSCheck.java [hunk 2]
```diff
@@ -90,24 +84,46 @@ private void checkMethod(Tree tree) {
...
     List<MethodInvocationTree> addAllowedOrigin = new ArrayList<>();
     List<MethodInvocationTree> applyPermit = new ArrayList<>();
 
     @Override
     public void visitMethodInvocation(MethodInvocationTree mit) {
       if (SET_ADD_HEADER_MATCHER.anyMatch(mit)) {
-        String constantCORS = ExpressionsHelper.getConstantValueAsString(mit.arguments().get(0)).value();
-        if (constantCORS != null && HTTP_HEADERS.contains(constantCORS.toLowerCase(Locale.ENGLISH))) {
+        String headerName = ExpressionsHelper.getConstantValueAsString(mit.arguments().get(0)).value();
+        if (ACCESS_CONTROL_ALLOW_ORIGIN.equalsIgnoreCase(headerName) && isStar(mit.arguments().get(1))) {
```
Comment (quentin-jaquier-sonarsource) [inline java-checks/src/main/java/org/sonar/java/checks/CORSCheck.java]: This is to support the case:
> HttpServletResponse: raise only when "Access-Control-Allow-Origin" header is set to "*"

File: java-checks/src/test/files/checks/CORSCheck.java [hunk 3]
```diff
@@ -1,53 +1,86 @@
...
       return ResponseEntity.ok().body("ok");
     }
+
+    @CrossOrigin // Noncompliant
+    @RequestMapping(value = "/test2")
+    public ResponseEntity<String> test2() {
+      return ResponseEntity.ok().body("ok");
+    }
+
+    @CrossOrigin("http://domain2.com") // Noncompliant
```
Comment (quentin-jaquier-sonarsource) [inline java-checks/src/test/files/checks/CORSCheck.java]: I juste realized that the key `value` is an alias for `origins`, this one, and a few following ones are in fact compliant!

File: java-checks/src/main/java/org/sonar/java/checks/CORSCheck.java [hunk 4]
```diff
@@ -90,24 +84,46 @@ private void checkMethod(Tree tree) {
...
+        !isStar(assignment.expression());
+    }
+    return false;
+  }
+
   private void reportTree(Tree tree) {
     reportIssue(tree, MESSAGE);
   }
 
+  private static boolean isStar(ExpressionTree expressionTree) {
```
Comment (alban-auzeill) [inline java-checks/src/main/java/org/sonar/java/checks/CORSCheck.java]: I would change the `isStar` method by `containsStar` rule to cover:
```java
    @CrossOrigin(origins = {"*"}) // Noncompliant
```
and
```java
    @CrossOrigin(origins = {"http://localhost:7777", "*"}) // Noncompliant
```

**Pull request merged**